proposals/features-contribute-lifecycle-scripts.md

```diff
@@ -0,0 +1,142 @@
+# [Proposal] Allow Dev Container Features to contribute lifecycle scripts
+
+Related to: https://github.com/devcontainers/spec/issues/60, https://github.com/devcontainers/spec/issues/181
+
+## Goal
+
+Allow Feature authors to provide lifecycle hooks for their Features during a dev container build.
+
+## Proposal
+
+ Introduce the following properties to [devcontainer-feature.json](https://containers.dev/implementors/features/#devcontainer-feature-json-properties), mirroring the behavior and syntax of the `devcontainer.json` lifecycle hooks.
+
+- "onCreateCommand"
+- "updateContentCommand"
+- "postCreateCommand"
+- "postStartCommand"
+- "postAttachCommand"
+
+Note that `initializeCommand` is omitted, pending further discussions around a secure design.
+
+Additionally, introduce a `${featureRoot}` environment variable, which is expanded at build time to the root of the expanded Feature the variable is used from.  
+
+Feature lifecycle hooks will be prepended to the lifecycle hooks (if any) declared by the current `devcontainer.json`. Lifecycle hooks will be prepended in the order that the Feature's installation script was executed.  Commands will be joined in such a way that a failure in any one lifecycle hook, will indicate a failure for that entire lifecycle hook execution.
+
+### Parallel Execution
+
+Any Features that declare lifecycle hooks using the [parallel execution syntax](https://containers.dev/implementors/spec/#parallel-exec) will be executed in parallel with any other Features or user-contributed lifecycle scripts. Any Features that do not use this syntax will continue to be run in sequence, blocking subsequent runs until the script exits successfully.
```

Comment:
Currently all scripts of a lifecycle hook are run in parallel. We can change that, but it seems a bit artificial to make this dependent on the syntax used.

The different hooks run sequentially, e.g., any "postStartCommand" script will only be run after all "postCreateCommand" scripts have finished. We probably want to keep it this way, the wording in the proposal should make this clearer. (The example in the linked doc needs to be updated to run the commands in the background, these "postCreateCommand" scripts will block later hooks from running.)

Reply:
All lifecycle hooks are run in parallel regardless of if the "parallel execution syntax" is used? That actually seems to line up with this [unit test](https://github.com/devcontainers/cli/compare/joshspicer/features-lifecycles-hooks#diff-05813acb3066795d8650780efe8bf09807c62a1cde94cf9d6b7ee0c67bdb73fbR48-R64) I wrote, where the execution of each script is not deterministic (which would make sense if they're running at the exact same time).

This was all documented based on my previous understanding of how this code worked - but I agree that ordering would be artificial and not something we should introduce with this change.

Reply:
I will collapse this spec down to indicate that scripts will indeed run in parallel.

Reply:
Although, I can think of instances where running _all_ lifecycle hooks in parallel may cause inconsistent behavior.  

For example, say two Features provide an `onCreateCommand` that both try to install an `apt` package.  The lock file will be taken by one script, and the other will fail to run. This behavior also has the change of being inconsistent, which we want to avoid with dev containers as much as possible.

Since our current offering for running commands in parallel is clearly opt-in (using the object syntax), there's less concern that users will accidentally get themselves stuck in a spot like this.  Additionally, Features obscure the contents of a script, so folks may not even realize there is a race condition in two Features' lifecycle scripts.

Does that make sense? I want us to discuss that before I move forward with serialization/parallelization of lifecycle hooks.

Reply:
Speaking as a feature author, I think my expectation would be that the scripts run in sequence, especially if the `installAfter` option was used. This is mainly because I think of the features as installing in a sequence, and so the scripts running in a similar sequence just seems like how it would work.

That said, running them in parallel feels fine and manageable. If I really need something else to finish first, it does not seem impossible to have the script I provided detect and wait for some condition to be satisfied .. and if I control the feature I need to wait for, I can even create some kind of marker file to make this easy.

Reply:
I'm fine with running all in parallel for now, but agree with the above that it would make sense to run scripts from different features in sequence to avoid conflicts. Scripts from the same feature could still run in parallel as we could expect each feature to manage dependencies among its scripts. If we run the scripts in sequence, it will make sense to do it in the same sequence as we ran the install scripts in case the hooks also have the same dependencies.

Reply:
I agree with what @chrmarti mentions above. 👍